src/components/sets/sets.js

```diff
@@ -511,8 +511,23 @@ export default class SetsTree {
    */
   newTab(setKey) {
     const node = this.findNode(setKey);
-    this.tabRoots = [...this.tabRoots, node];
-    this.emitTreeUpdate();
+    // Only add a tab if it does not already exist.
+    if (!this.tabRoots.find(t => t.setKey === setKey)) {
+      this.tabRoots = [...this.tabRoots, node];
+      this.emitTreeUpdate();
+    }
+  }
+
+  /**
+   * Remove a tab root by its key.
+   * @param {string} setKey The key of the tab root node to be removed.
+   * @param {boolean} preventEmit Whether to prevent the emit event.
+   */
+  closeTab(setKey, preventEmit) {
+    this.tabRoots = this.tabRoots.reduce((a, h) => (h.setKey === setKey ? a : [...a, h]), []);
```

Comment:
Almost the same thing on line 425: Could you factor out a function, and explain what it does?